WalletWasabi/Blockchain/Keys/KeyManager.cs

```diff
@@ -274,88 +309,23 @@ public void SetFilePath(string? filePath)
 		IoHelpers.EnsureContainingDirectoryExists(FilePath);
 	}
 
-	public void ToFile()
-	{
-		lock (HdPubKeysLock)
-		{
-			lock (BlockchainStateLock)
-			{
-				lock (ToFileLock)
-				{
-					ToFileNoLock();
-				}
-			}
-		}
-	}
-
-	public void ToFile(string filePath)
-	{
-		lock (HdPubKeysLock)
-		{
-			lock (BlockchainStateLock)
-			{
-				lock (ToFileLock)
-				{
-					ToFileNoLock(filePath);
-				}
-			}
-		}
-	}
-
 	public HdPubKey GenerateNewPersistentKey(SmartLabel label, KeyState keyState, bool isInternal)
 	{
 		var newKey = GenerateNewKey(label, keyState, isInternal);
 		ToFile();
 		return newKey;
 	}
 
-	public HdPubKey GenerateNewKey(SmartLabel label, KeyState keyState, bool isInternal)
+	public HdPubKey GenerateNewKey(SmartLabel label, KeyState keyState, bool isInternal, ScriptPubKeyType scriptPubKeyType = ScriptPubKeyType.Segwit)
 	{
-		// BIP44-ish derivation scheme
-		// m / purpose' / coin_type' / account' / change / address_index
-		var change = isInternal ? 1 : 0;
+		var hdPubKeyRegistry = GetHdPubKeyGenerator(isInternal, scriptPubKeyType);
 
-		lock (HdPubKeysLock)
+		lock (CriticalStateLock)
 		{
-			HdPubKey[] relevantHdPubKeys = HdPubKeys.Where(x => x.IsInternal == isInternal).ToArray();
-
-			KeyPath path = new($"{change}/0");
-			if (relevantHdPubKeys.Any())
-			{
-				int largestIndex = relevantHdPubKeys.Max(x => x.Index);
-				var smallestMissingIndex = largestIndex;
-				var present = new bool[largestIndex + 1];
-				for (int i = 0; i < relevantHdPubKeys.Length; ++i)
-				{
-					present[relevantHdPubKeys[i].Index] = true;
-				}
-				for (int i = 1; i < present.Length; ++i)
-				{
-					if (!present[i])
-					{
-						smallestMissingIndex = i - 1;
-						break;
-					}
-				}
-
-				path = relevantHdPubKeys[smallestMissingIndex].NonHardenedKeyPath.Increment();
-			}
-
-			var fullPath = AccountKeyPath.Derive(path);
-			var pubKey = ExtPubKey.Derive(path).PubKey;
-
-			var hdPubKey = new HdPubKey(pubKey, fullPath, label, keyState);
-			HdPubKeys.Add(hdPubKey);
-			lock (HdPubKeyScriptBytesLock)
-			{
-				HdPubKeyScriptBytes.Add(hdPubKey.P2wpkhScript.ToCompressedBytes());
-			}
-
-			lock (ScriptHdPubKeyMapLock)
-			{
-				ScriptHdPubKeyMap.Add(hdPubKey.P2wpkhScript, hdPubKey);
-			}
-
+			var view = HdPubKeyCache.GetView(hdPubKeyRegistry.KeyPath); 
```

Comment:
It seems that methods is for testing purposes only and that's why i didn't see any problem. Will change it as internal and raise exception anyway in case of null.

Reply:
Done